docs/WASI-design-principles.md

```diff
@@ -0,0 +1,93 @@
+
+Note that this is a different sense of "capability" than [Linux capabilities](http://man7.org/linux/man-pages/man7/capabilities.7.html) or the withdrawn [POSIX capabilities](https://archive.org/details/posix_1003.1e-990310), which are per-process rather than per-resource.
+
+The simplest representation of handles are values of [reference type](https://github.com/WebAssembly/reference-types). References in wasm are inherently unforgeable, so they can represent handles directly.
+
+Some programming languages operate primarily within linear memory, such as C, C++, and Rust, and there currently is no easy way for these languages to use references in normal code. And even if it does become possible, it's likely that source code will still require annotations to fully opt into references, so it won't always be feasible to use. For these languages, references are stored in a [table](https://webassembly.github.io/spec/core/bikeshed/index.html#tables%E2%91%A0) called a *c-list*. Integer indices into the table then identify resources, which can be easily passed around or stored in memory. In some contexts, these indices are called *file descriptors* since they're similar to what POSIX uses that term for. There are even some tools, such as wasm-bindgen, which make this fairly easy. (Internally, tools and engines don't always use actual WebAssembly tables to do this, however those are implementation details. Conceptually, they work as if they had tables.)
+
+If all instances shared a single c-list, that would undermine isolation, because any instance could forge an index to any other instance's resources. Fortunately, there can be multiple c-lists, and instances can be given access to some c-lists and not others, or even have instances which don't use any c-lists at all, so it's still possible to establish isolation between instances.
+
+Witx-specified APIs use a special `handle` keyword to mark parameters and return values which are handles. In the short term, these are lowered to integer indices, with an implied table, so that the APIs can be easily used from C and similar languages today. Once [interface types](https://github.com/WebAssembly/interface-types) and [type imports](http://github.com/WebAssembly/type-imports-proposal) are ready, we expect to make use of them to provide APIs which can be used either from languages using references or from languages using integer indices, with tables being used and managed automatically.
```

Comment:
noticed a broken link, should be https://github.com/WebAssembly/proposal-type-imports